### PR TITLE
New CLI for AMICAL

### PR DESCRIPTION
**Pull request by DrSoulain**
We start to implement the Command Line Interface (CLI) for AMICAL. The idea is to use AMICAL both using the scripted method (e.g., doc/example_NIRISS.py) and directly in the prompt. For now, the 3 main features are implemented: 
- clean (clean the post-pipeline data to be NRM compliant),
- extract (extract the raw observable saved as pickled object),
- calibrate (calibrate and save them as oifits file).

### CODE REVIEW

File: amical/mf_pipeline/ami_function.py [hunk 1]
```diff
@@ -400,6 +404,8 @@ def make_mf(
 
     if display:
         _plot_mask_coord(xy_coords, maskname, instrument)
+        if save:
+            plt.savefig(figdir + Path(filename).stem + "_1.pdf")
```
Comment (neutrinoceros) [inline amical/mf_pipeline/ami_function.py:408]: This can't work if figdir is left to it's default value (`None`)
I suggest refactoring this so that we have a single argument `save_to` (or `save_dir`, up to you) that replaces `figdir` and `save`

File: amical/mf_pipeline/ami_function.py [hunk 2]
```diff
@@ -554,7 +560,7 @@ def make_mf(
     im_uv = np.roll(np.fft.fftshift(mf_tot), 1, axis=1)
 
     if display:
-        plt.figure(figsize=(6, 6))
+        plt.figure(figsize=(9, 7))
```
Comment (neutrinoceros) [inline amical/mf_pipeline/ami_function.py:563]: is this change really needed for this PR ?

File: amical/mf_pipeline/bispect.py [hunk 3]
```diff
@@ -686,6 +688,7 @@ def _normalize_all_obs(
     index_mask,
     infos,
     expert_plot=False,
+    save=False,
```
Comment (neutrinoceros) [inline amical/mf_pipeline/bispect.py:691]: same as above
```suggestion
    *
    save_to=None,
```

File: amical/mf_pipeline/bispect.py [hunk 4]
```diff
@@ -1016,6 +1019,20 @@ def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix):
     return infos
 
 
+def produce_result_pdf(figdir, filename):
+    # Call the PdfFileMerger
+    mergedObject = PdfFileMerger()
+
+    for fileNumber in range(7):
+        ifile = figdir + filename + "_" + str(fileNumber + 1) + ".pdf"
```
Comment (neutrinoceros) [inline amical/mf_pipeline/bispect.py:1027]: This isn't really robust
```suggestion
        ifile = os.path.join(figdir, f"{filename}_{fileNumber + 1}.pdf")
```

File: amical/mf_pipeline/bispect.py [hunk 5]
```diff
@@ -1016,6 +1019,20 @@ def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix):
...
+        mergedObject.append(PdfFileReader(ifile, "rb"))
+        os.remove(ifile)
+
+    # Write all the files into a file which is named as shown below
+    mergedObject.write(figdir + filename + "_DIAGNOSTIC_PLOTS.pdf")
```
Comment (neutrinoceros) [inline amical/mf_pipeline/bispect.py:1032]:
```suggestion
    mergedObject.write(os.path.join(figdir, f"{filename}_DIAGNOSTIC_PLOTS.pdf")
```

File: amical/mf_pipeline/bispect.py [hunk 6]
```diff
@@ -1037,6 +1054,7 @@ def extract_bs(
     unbias_v2=True,
     compute_cp_cov=True,
     expert_plot=False,
+    save=False,
```
Comment (neutrinoceros) [inline amical/mf_pipeline/bispect.py:1057]: this looks like an early merge from a different PR ?
In any case, new arguments should be appended at the end of the signature.
Also, new arguments, especially boolean, should be made keyword only

File: amical/mf_pipeline/bispect.py [hunk 7]
```diff
@@ -1105,6 +1123,11 @@ def extract_bs(
         cprint("\n-- Starting extraction of observables --", "cyan")
     start_time = time.time()
 
+    figdir = "amical_save_fig/"
+    if save:
```
Comment (neutrinoceros) [inline amical/mf_pipeline/bispect.py:1127]: see my comment above about having a single argument instead of two here

File: amical/mf_pipeline/bispect.py [hunk 8]
```diff
@@ -1151,7 +1174,17 @@ def extract_bs(
...
         display=display,
+        save=save,
+        figdir=figdir,
+        filename=filename,
     )
+
+    figname = figdir + Path(filename).stem
+    ifig = 2
+    if save:
+        plt.savefig(figname + "_%i.pdf" % ifig)
```
Comment (neutrinoceros) [inline amical/mf_pipeline/bispect.py:1185]:
```suggestion
    filename, _ = os.path.splitext(filename)
    figname = os.path.join(figdir, filename)
    ifig = 2
    if save:
        plt.savefig(f"{figname}_{ifig}.pdf")
```

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
here are a couple of comments to start with. Please carefully review all places where you're constructing a path to a directory or a file and make sure you use either `os.path` or `pathlib` to to so (avoid string concatenation with `+`, this is a very easy way to write bugs that can be avoided)

File: amical/_cli/main.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    args = parser.parse_args(argv)
+
+    if args.command == "clean":
+        perform_clean(args)
+    elif args.command == "extract":
+        perform_extract(args)
+    elif args.command == "calibrate":
+        perform_calibrate(args)
+
+    return 0
```
Comment (neutrinoceros) [inline amical/_cli/main.py:334]: So as to make the return value any useful, I suggest refactoring this so that command functions return integers to, and eventually cascades _their_ return value into main as
```suggestion
    retv = 1
    if args.command == "clean":
        retv = perform_clean(args)
    elif args.command == "extract":
        retv = perform_extract(args)
    elif args.command == "calibrate":
        retv = perform_calibrate(args)
    return retv
```

File: amical/mf_pipeline/ami_function.py [hunk 10]
```diff
@@ -327,6 +328,9 @@ def make_mf(
     diag_plot=False,
     verbose=False,
     display=True,
+    save=False,
+    figdir=None,
+    filename=None,
```
[the remark this answers is not in the document]
Comment (DrSoulain) [reply amical/mf_pipeline/ami_function.py:333]: What did you add here?

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply amical/mf_pipeline/ami_function.py:333]: `*,`, it makes following arguments keyword-only, see https://www.python.org/dev/peps/pep-3102/

File: amical/_cli/api.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    d = []
+    for i, f in enumerate(l_file):
+        bs = amical.load_bs_hdf5(f)
+        filename = Path(f).stem
+        hdr = bs.infos.hdr
+        target = hdr.get("OBJECT", None)
+        date = hdr.get("DATE-OBS", None)
+        ins = hdr.get("INSTRUME", None)
+
+        if target is not None or "":
```
Comment (neutrinoceros) [inline amical/_cli/api.py:77]:
```suggestion
        if target not in (None or ""):
```

File: amical/_cli/api.py [hunk 12]
```diff
@@ -74,7 +74,7 @@ def _select_association_file(args):
         date = hdr.get("DATE-OBS", None)
         ins = hdr.get("INSTRUME", None)
 
-        if target is not None or "":
+        if target not in (None or ""):
```
Comment (neutrinoceros) [inline amical/_cli/api.py:77]: Oops, sorry about that
```suggestion
        if target not in (None, ""):
```

File: amical/_cli/api.py [hunk 13]
```diff
@@ -74,13 +74,13 @@ def _select_association_file(args):
         date = hdr.get("DATE-OBS", None)
         ins = hdr.get("INSTRUME", None)
 
-        if target not in (None or ""):
+        if (target is not None) & (target != "Unknown"):
```
Comment (neutrinoceros) [inline amical/_cli/api.py:77]: `&`  isn't short for `and`, it's a different operator
```suggestion
        if (target is not None) and (target != "Unknown"):
```

File: amical/tests/test_cli.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            "--datadir",
+            str(cli_datadir),
+            "--outdir",
+            str(tmp_path),
+            "--isz",
+            str(isz),
+        ]
+    )
+
+    input_file = sorted(glob(str(cli_datadir) + "/*.fits"))
```
Comment (neutrinoceros) [inline amical/tests/test_cli.py:31]:
```suggestion
    input_file = sorted(cli_datadir.glob("*.fits"))
```

File: amical/tests/test_cli.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+            ]
+        )
+
+    for i in range(2):
+        monkeypatch.setattr("builtins.input", lambda _: str(i))
+        main(["extract", "--datadir", str(tmp_path), "--outdir", str(tmp_path)])
+
+    monkeypatch.setattr("builtins.input", lambda _: "1")
+    monkeypatch.setattr("builtins.input", lambda _: "0")
+
```
Comment (neutrinoceros) [inline amical/tests/test_cli.py:125]:
```suggestion
```

File: amical/tests/test_cli.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+                "--outdir",
+                str(tmp_path),
+                "--peakmethod",
+                method,
+            ]
+        )
+
+    monkeypatch.setattr("builtins.input", lambda _: "1")
+    monkeypatch.setattr("builtins.input", lambda _: "0")
+
```
Comment (neutrinoceros) [inline amical/tests/test_cli.py:182]:
```suggestion
```

File: amical/tests/test_cli.py [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+@pytest.mark.parametrize("method", ["fft", "gauss", "unique", "square"])
+def test_calibrate_method(method, cli_datadir, tmp_path, monkeypatch):
+    plt.close("all")
+    monkeypatch.setattr("builtins.input", lambda _: "0")
+
+    for i in range(2):
+        monkeypatch.setattr("builtins.input", lambda _: str(i))
+
+        isz = 78
+        main(
```
Comment (neutrinoceros) [inline amical/tests/test_cli.py:154]:
```suggestion
        ret = main(
```

**Review by neutrinoceros — COMMENTED** [review]
Here are a couple suggestions on how to improve the test module, but otherwise it looks pretty good

https://pypi.org/project/pytest-randomly/

```
@pytest.fixture()
def close_figures():
    plt.close("all")
    yield
    plt.close("all")
 ```
 
 https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob

File: amical/_cli/commands/calibrate.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    d = []
+    for i, f in enumerate(l_file):
+        bs = amical.load_bs_hdf5(f)
+        filename = Path(f).stem
+        hdr = bs.infos.hdr
+        target = hdr.get("OBJECT", None)
+        date = hdr.get("DATE-OBS", None)
+        ins = hdr.get("INSTRUME", None)
+
+        if (target is not None) & (target != "Unknown"):
```
Comment (neutrinoceros) [inline amical/_cli/commands/calibrate.py:47]:
```suggestion
        if target not in (None, "Unknown"):
```
